crates/conjure_core/src/rules/normalisers/associative_commutative.rs

```diff
@@ -1,4 +1,50 @@
 //! Generic normalising rules for associative-commutative operators.
 
-// TODO: this file will contain a generic `normalise_associative_commutative` rule that replaces
-// `flatten_nested_sum`, `unwrap_nested_or`, and so on. See #491.
+use conjure_core::ast::Expression as Expr;
+use conjure_core::rule_engine::{
+    register_rule, ApplicationError::RuleNotApplicable, ApplicationResult, Reduction,
+};
+use conjure_core::Model;
+use uniplate::{Biplate, Uniplate};
+
+/// Normalises associative_commutative operations.
+///
+/// For now, this just removes nested expressions by associativity.
+///
+/// ```text
+/// v(v(a,b,...),c,d,...) ~> v(a,b,c,d)
+/// where v is an AC vector operator
+/// ```
+#[register_rule(("Base", 8400))]
+fn normalise_associative_commutative(expr: &Expr, _: &Model) -> ApplicationResult {
+    if !expr.is_associative_commutative_operator() {
+        return Err(RuleNotApplicable);
+    }
+
+    let children = <_ as Biplate<Vec<Expr>>>::children_bi(expr);
+
+    // empty expression
+    if children.is_empty() {
+        return Err(RuleNotApplicable);
+    }
+
+    let children = children[0].clone();
```

Comment:
I think we can do better.

1) Instead of special casing the first child, iterate over the children so ve support `v(v(...), v(...), ....)`

2) Instead of going one level deep, we can also handle `v(v(v(), _), _)` while we are at it.

Add a helper function that is a bit like children, but recursively calls itself on any child that has the same discriminant and concatanates the results. Can be called `associative_children`, or `nested_children`, or `deep_children`.

I think it's a small change and will make the rule much more generic.

Reply:
Trying it in Haskell to make sure it is implementable.

First, this wouldn't work (because you'd have to implement a seperate version per Op.)

```
deep_children (Op xs) = concatMap deep_children xs
deep_children x = [x]
```


Second, this also wouldn't work (because OpX's and OpY's would then get mixed up.)

```
deep_children (OpX xs) = concatMap deep_children xs
deep_children (OpY xs) = concatMap deep_children xs
deep_children x = [x]
```

This would work, I think: 

```
deep_children root@(OpX _) (OpX xs) = concatMap (deep_children root) xs
deep_children root x = [x]
```

i.e. we keep the root as a reference of which discriminant we are working on and pass it down.

You can instead have `root = std::mem::discriminant(expr)` in Rust, may be clearer.

I am still thinking in Haskell, but I think this can be one-to-one mapped to Rust.

Reply:
oops, no we won't. that part is fine then, only the nesting.

Reply:
> Trying it in Haskell to make sure it is implementable.
> 
> First, this wouldn't work (because you'd have to implement a seperate version per Op.)
> 
> ```
> deep_children (Op xs) = concatMap deep_children xs
> deep_children x = [x]
> ```
> 
> 
> Second, this also wouldn't work (because OpX's and OpY's would then get mixed up.)
> 
> ```
> deep_children (OpX xs) = concatMap deep_children xs
> deep_children (OpY xs) = concatMap deep_children xs
> deep_children x = [x]
> ```
> 
> This would work, I think: 
> 
> ```
> deep_children root@(OpX _) (OpX xs) = concatMap (deep_children root) xs
> deep_children root x = [x]
> ```
> 
> i.e. we keep the root as a reference of which discriminant we are working on and pass it down.
> 
> You can instead have `root = std::mem::discriminant(expr)` in Rust, may be clearer.
> 
> I am still thinking in Haskell, but I think this can be one-to-one mapped to Rust.
> 

 we could maybe do this as a transform / a transform_bi?

Reply:
to clarify: we might have operators that have multiple vectors, but not AC ones.

Reply:
>  we could maybe do this as a transform / a transform_bi?
Or, if we want to prune things a bit more, a (top down) descend

Reply:
I leave this in your capable hands :)